keyboards/moon/matrix.c

```diff
@@ -0,0 +1,202 @@
+
+__attribute__((weak)) void matrix_scan_kb(void) { matrix_scan_user(); }
+
+__attribute__((weak)) void matrix_init_user(void) {}
+
+__attribute__((weak)) void matrix_scan_user(void) {}
+
+inline uint8_t matrix_rows(void) { return MATRIX_ROWS; }
+
+inline uint8_t matrix_cols(void) { return MATRIX_COLS; }
+
+inline bool matrix_is_on(uint8_t row, uint8_t col) { return (matrix[row] & ((matrix_row_t)1 << col)); }
+
+inline matrix_row_t matrix_get_row(uint8_t row) { return matrix[row]; }
+
+void matrix_print(void) {
+  print_matrix_header();
+
+  for (uint8_t row = 0; row < MATRIX_ROWS; row++) {
+    phex(row);
+    print(": ");
+    print_matrix_row(row);
+    print("\n");
+  }
+}
+
+uint8_t matrix_key_count(void) {
+  uint8_t count = 0;
+  for (uint8_t i = 0; i < MATRIX_ROWS; i++) {
+    count += matrix_bitpop(i);
+  }
+  return count;
+}
+
+static void init_i2c(void) {
+  pca9555_init(IC1);
+  pca9555_init(IC2);
+}
+
+static void init_pins(void) {
+  // init cols - IC2 port0 & IC2 port1 input
+  pca9555_set_config(IC2, PCA9555_PORT0, ALL_INPUT);
+  pca9555_set_config(IC2, PCA9555_PORT1, ALL_INPUT);
+
+  // init rows - IC1 port0 output
+  pca9555_set_config(IC1, PCA9555_PORT0, ALL_OUTPUT);
+  pca9555_set_output(IC1, PCA9555_PORT0, ALL_HIGH);
+}
+
+static void select_row(uint8_t row) {
+  // All rows are on the same IC and port
+  uint8_t mask = 1 << row;
+
+  // set active row low  : 0
+  // set other rows hi-Z : 1
+  pca9555_set_output(IC1, PCA9555_PORT0, ALL_HIGH & (~mask));
+}
+
+static uint32_t read_cols(void) {
+  uint32_t state_1 = pca9555_readPins(IC2, PCA9555_PORT0);
+  uint32_t state_2 = pca9555_readPins(IC2, PCA9555_PORT1);
+
+  uint32_t state = ((state_1 & PORT0_COLS_MASK) << 3) | ((state_2 & PORT1_COLS_MASK));
+
+  // A low pin indicates an active column
+  return (~state) & COLS_MASK;
+}
+
+static bool read_cols_on_row(matrix_row_t current_matrix[], uint8_t current_row) {
+  // Store last value of row prior to reading
+  matrix_row_t last_row_value = current_matrix[current_row];
+
+  // Clear data in matrix row
+  current_matrix[current_row] = 0;
+
+  // Select row and wait for row selecton to stabilize
+  select_row(current_row);
+  wait_us(30);
+
+  current_matrix[current_row] |= read_cols();
+
+  // No need to unselect as `select_row` sets all the pins.
+
+  return (last_row_value != current_matrix[current_row]);
+}
+
+void matrix_init(void) {
+  // initialize i2c
+  init_i2c();
+
+  // initialize key pins
+  init_pins();
+
+  // initialize matrix state: all keys off
+  for (uint8_t i = 0; i < MATRIX_ROWS; i++) {
+    raw_matrix[i] = 0;
+    matrix[i]     = 0;
+  }
+
+  debounce_init(MATRIX_ROWS);
+
+  matrix_init_quantum();
+}
+
+uint8_t matrix_scan(void) {
+  bool changed = false;
+
+  for (uint8_t current_row = 0; current_row < MATRIX_ROWS; current_row++) {
+    changed |= read_cols_on_row(raw_matrix, current_row);
+  }
+
+  debounce(raw_matrix, matrix, MATRIX_ROWS, changed);
+
+  matrix_scan_quantum();
+
+  return 1;
```

Comment:
A few features rely on this to be set correctly. 

To keep things consistent, could you change it to: 
```suggestion
  return (uint8_t)changed;
```

Reply:
Fixed. Though the initial code came from the `quantum/matrix.c` file.

Reply:
Yeah.  There is a PR that fixes the behavior, that was opened today because of issues with the matrix.  So, since it will be updated very soon... it's easier to get this file fixed now, than to wait for later to fix it! 

#5984 is the PR in question

Reply:
I see, I wasn't aware of that PR. I agree wholeheartedly that it is better to fix it right away.